ark/jsonschema/array.ts

```diff
@@ -0,0 +1,159 @@
+		else seen[stringified] = true
+	}
+	return duplicates.length === 0 ?
+			true
+		:	ctx.reject({
+				expected: "unique array items",
+				actual: `duplicated at elements ${printable(duplicates)}`
+			})
+}
+
+const arrayContainsItemMatchingSchema = (
+	array: readonly unknown[],
+	schema: Type<unknown>,
+	ctx: TraversalContext
+) =>
+	array.some(item => schema.allows(item)) === true ?
+		true
+	:	ctx.mustBe(
+			"an array containing at least one item matching 'contains' schema"
+		)
+
+export const validateJsonSchemaArray = JsonSchema.ArraySchema.pipe(
+	jsonSchema => {
+		const arktypeArraySchema: Intersection.Schema<Array<unknown>> = {
+			proto: "Array"
+		}
+
+		if ("items" in jsonSchema) {
+			if (Array.isArray(jsonSchema.items)) {
+				arktypeArraySchema.sequence = {
+					prefix: jsonSchema.items.map(
+						item => innerParseJsonSchema.assert(item).internal
+					)
+				}
+
+				if ("additionalItems" in jsonSchema) {
+					if (jsonSchema.additionalItems === false)
+						arktypeArraySchema.exactLength = jsonSchema.items.length
+					else {
+						arktypeArraySchema.sequence = {
+							...arktypeArraySchema.sequence,
+							variadic: innerParseJsonSchema.assert(jsonSchema.additionalItems)
+								.internal
+						}
+					}
+				}
+			} else {
+				arktypeArraySchema.sequence = {
+					variadic: innerParseJsonSchema.assert(jsonSchema.items).internal
+				}
+			}
+		}
+
+		if ("maxItems" in jsonSchema)
+			arktypeArraySchema.maxLength = jsonSchema.maxItems
+		if ("minItems" in jsonSchema)
+			arktypeArraySchema.minLength = jsonSchema.minItems
+
+		const predicates: Predicate.Schema[] = []
+		if ("uniqueItems" in jsonSchema && jsonSchema.uniqueItems === true)
+			predicates.push((arr: unknown[], ctx) => arrayItemsAreUnique(arr, ctx))
+
+		if ("contains" in jsonSchema) {
+			const parsedContainsJsonSchema = innerParseJsonSchema.assert(
+				jsonSchema.contains
+			)
+			predicates.push((arr: unknown[], ctx) =>
+				arrayContainsItemMatchingSchema(arr, parsedContainsJsonSchema, ctx)
+			)
+		}
+
+		arktypeArraySchema.predicate = predicates
+
+		return rootSchema(arktypeArraySchema) as unknown as Type<JsonSchema.Json[]>
+	}
+)
+
+type inferArrayOfJsonSchema<tuple extends array<JsonSchema.Schema>> = {
+	[index in keyof tuple]: inferJsonSchema<tuple[index]>
+}
+
+export type inferJsonSchemaArray<arraySchema, T = unknown> =
+	"additionalItems" extends keyof arraySchema ?
+		"items" extends keyof arraySchema ?
+			arraySchema["items"] extends array<JsonSchema.Schema> ?
+				inferJsonSchemaArrayConstraints<
+					Omit<arraySchema, "additionalItems" | "items">,
+					// @ts-ignore - TypeScript complains that this is "excessively deep", despite it correctly resolving the type
+					[
+						...inferJsonSchemaArrayItems<arraySchema["items"]>,
+						...inferJsonSchema<arraySchema["additionalItems"]>[]
+					]
+				>
+			:	// JSON Schema spec explicitly says that additionalItems MUST be ignored if items is not an array, and it's NOT an error
```

Comment:
We don't have an obligation to support everything that's technically valid JSON schema if it seems like a clear mistake.

I'd rather throw here

Reply:
If it was a "should" then I'd agree, but because it's "must" I'd prefer to keep it as-is